### PR TITLE
docs(blueprint): reasons not to generate openAPI schema from API calls

### PR DESCRIPTION
**Pull request by JammUtkarsh**
## Related Issue
  - This PR can used as platform to discuss the idea proposed.

Closes: #432

#### Describe the changes you've made
A `blueprint/docs` directory has been added. It could be used to record all architectural discussions made in while keploy.

We could use the code review(line by line) in order to discuss on each point if needed.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Please let us know if any test cases are added

No tests or test cases added.

#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)
NA

## Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.

## Screenshots (if any)

|        Original         |          Updated           |
|:-----------------------:|:--------------------------:|
| **original screenshot** | <b>updated screenshot </b> |

### CODE REVIEW

**Comment by JammUtkarsh** [comment]
I have read the CLA Document and I hereby sign the CLA

File: blueprint/proposal/openAPISchemaFromAPICalls.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+- An API tester who might have aimed to generate schema may miss on varios edge cases, various data type combinations, etc. They might not be aware of all the possible scenarios and will be limted to the API calls they have made and experience they have had.
+- The openAPI schema has a `maximum` and `minimum` field which can be used to define the upper and lower limits of a field value. This is not possible to be defined in an mock API call.
+- No defination upper and lower limits for fields. This is not possible to be defined in the API call.  For example, a rating system can have a rating between 1 and 5 but this can not be defined in the API call.
+
+These are some of the problems that I can think of. There might be more problems that I am not aware of.
+
+## Conclusion
+
+I think the openAPI schema should be the source of truth for the API. The openAPI schema should be generated in the proper stage of software development life cycle. The openAPI schema should be generated by the API developer and not by the API tester. The API tester should use the openAPI schema to generate test cases.
```
Comment (charankamarapu) [inline blueprint/proposal/openAPISchemaFromAPICalls.md:30]: Like I have commented earlier it is not about who does it tester/developer majorly we need a less human interactive way to produce openAPI schema.

File: blueprint/proposal/openAPISchemaFromAPICalls.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+Repository: <https://github.com/keploy/keploy/>
+
+Issue: [#432](https://github.com/keploy/keploy/issues/432)
+
+---
+
+## Overview
+
+The idea proposed in the issue is to be able to generate test cases from API calls. This is certainly a good idea, but if it would have been possible then it would have been done already and used extensively. The reason  openAPI specification exists because the schema is the source of truth for the API.
```
[the remark this answers is not in the document]
Comment (JammUtkarsh) [reply blueprint/proposal/openAPISchemaFromAPICalls.md:11]: Yes, thats true. I will fix the statement in the next commit.

File: blueprint/proposal/openAPISchemaFromAPICalls.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+Also, the task of writing schema is done in the planning phase of the project. It is not something that is done after the project is completed. So, the openAPI spec should already be available before the project is completed. I don't think following a non-conventional reverse approach might be a good idea.
```
[the remark this answers is not in the document]
Comment (JammUtkarsh) [reply blueprint/proposal/openAPISchemaFromAPICalls.md:13]: >But most of the developers do not write in openAPI format they just write in a verbal format where they could explain it to others.

In my opinon, I don't think that's the right approch. Unless it's a side project one can skip generating it.

>To produce testCases we need a proper openAPI format schema that is the reason this task is created.

We can take this discussion in another PR which is relevent to the issue [#338](https://github.com/keploy/keploy/issues/338).

From a developer's standpoint, I think I would choose specilized tool like [Swagger Inspector](https://inspector.swagger.io/builder) to produce the basic API schema; [Swagger Editor](https://editor.swagger.io/) to build the rest of work; and finally when schema is ready, I will ingest it in Keploy(future release) test my API against it.

TL;DR: If Swagger specilises in schema generation and Keploy in testing API against schema, then I would use them respectively.

File: blueprint/proposal/openAPISchemaFromAPICalls.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+- At an abstract level the openAPI schema is a contract between the client and the server. The client is expected to send the request in a certain format and the server is expected to send the response in a certain format. If the API call does not define the format of the request and response, then the openAPI schema might not be valid.
+- An API tester who might have aimed to generate schema may miss on varios edge cases, various data type combinations, etc. They might not be aware of all the possible scenarios and will be limted to the API calls they have made and experience they have had.
+- The openAPI schema has a `maximum` and `minimum` field which can be used to define the upper and lower limits of a field value. This is not possible to be defined in an mock API call.
+- No defination upper and lower limits for fields. This is not possible to be defined in the API call.  For example, a rating system can have a rating between 1 and 5 but this can not be defined in the API call.
+
+These are some of the problems that I can think of. There might be more problems that I am not aware of.
```
[the remark this answers is not in the document]
Comment (JammUtkarsh) [reply blueprint/proposal/openAPISchemaFromAPICalls.md:26]: >Yes you are correct regarding the problems that may arise due to generation of OpenAPI using API calls. But the major agenda of this task is to generate the OpenAPI schema automatically with not much human interaction or work .

Like I said in the above comment.

>If you can think of anyway to produce OpenAPI schema using may be AI or some tools or libraries which parse through code and produce it or any other way

1. Using an AI model could be expensive both from user's standpoint(computational power spent on running the model, if we built it) and from Keploy's perspective(running the infrastructure to run the model)
2. Like I mentioned above some tools are [Swagger Inspector](https://inspector.swagger.io/builder) and [Swagger UI](https://editor.swagger.io/).

**Comment by slayerjain** [comment]
@charankamarapu for answering the questions. 

@JammUtkarsh In summary, Keploy's purpose is test generation. We would need to serve broad category of developers without enforcing too many conditions. Great tools are aware that best practices may not be followed but handle those situation gracefully.

Since we are creating a test generator which takes input schema, it would benefit to have a schema generator too (unless there's something readily available and general purpose) - so that we can serve applications which do not have a schema.

File: blueprint/proposal/openAPISchemaFromAPICalls.md (hunk 1 above)
Comment (PranshuSrivastava) [reply blueprint/proposal/openAPISchemaFromAPICalls.md:30]: @JammUtkarsh The main motivation behind this project is to make using Keploy a hands free experience. If the user already has an open api schema we will use it, and if they don't, we want to be able to generate the schema on our own so that they don't have to do it. If the developer does not like this behaviour,  they are free to generate their own schema using the tools you mentioned, and then give it to Keploy. But we want the choice to remain with the user.
As far as the problems are concerned, we will have to try and solve them as we move along in the project.

**Comment by JammUtkarsh** [comment]
@slayerjain | @charankamarapu | @PranshuSrivastava 

Thank you guys, now I have a better understanding of the purpose of this issue/feature.

If I'll be assigned with this task then, I would prefer to do it on a later stage. Since I would want to first learn
1. How to parse schema and convert them to go data structures?
2. How to generate tests with data structures in step 1?
3. Estimate computational cost(as in how much time and resources are required).

This would give me a better understanding of dynamics of API and how effectively the tests are being generated. Then go forward to generate schema from API calls. 

Moving from controlled and less chaotic enviroment to a choatic and dynamic environment.

The maintainers can feel free to close the PR.

**Comment by PranshuSrivastava** [comment]
@JammUtkarsh Sure, if you feel like this project is not for you right now, you can take up some other task. However, this an evaluation task for the [4th gsoc project](https://github.com/keploy/gsoc/tree/main/2023#4-autogenerate-test-cases). If you wish to apply for that project, then you will need to submit a PR for this as we will be selecting mentees based on the quality of their PRs for the mentioned tasks.

**Comment by charankamarapu** [comment]
This PR is sort of discussion rather than solution for an issue . So closing the PR.
